allennlp/nn/util.py

```diff
@@ -12,7 +12,12 @@
 import numpy
 import torch
 import torch.distributed as dist
-from torch.distributed import ReduceOp
+
+if dist.is_available():
+    from torch.distributed import ReduceOp
+else:
+    class ReduceOp(object):
+        SUM = None
```

Comment:
`ReduceOp` is actually only imported for use a type hint. So we could do this instead:

```suggestion
if TYPE_CHECKING:
    from torch.distributed import ReduceOp
```

The `TYPE_CHECKING` variable can be imported on line 9 above from the `typing` module.

Reply:
Sorry, but how do we rewrite `def dist_reduce(value: _V, reduce_op: ReduceOp, **kwargs)` when `TYPE_CHECKING` is False?

Reply:
And I found that [average.py](https://github.com/allenai/allennlp/blob/main/allennlp/training/metrics/average.py#L4) imported `ReduceOp`. Is it really "type hint" use?

Reply:
In `def dist_reduce(value: _V, reduce_op: ReduceOp, **kwargs)`, `ReduceOp` is simply a type hint. The Python interpreter doesn't care about type hints. And since the `TYPE_CHECKING` flag is `False` when the interpreter is running, you can use that flag to exclude imports that are only used as type hints.

The `TYPE_CHECKING` flag is only true when you're running `mypy` or some other type-checking program.

Reply:
I've tried your suggestion above, but it causes another error:

```
>>> import allennlp.training.metrics.average
  File "<stdin>", line 1, in <module>
  File "/home/yasuoka/.local/lib/python3.7/site-packages/allennlp/training/__init__.py", line 1, in <module>
    from allennlp.training.checkpointer import Checkpointer
  File "/home/yasuoka/.local/lib/python3.7/site-packages/allennlp/training/checkpointer.py", line 13, in <module>
  File "/home/yasuoka/.local/lib/python3.7/site-packages/allennlp/nn/util.py", line 2024, in <module> 
    def dist_reduce(value: _V, reduce_op: ReduceOp, **kwargs) -> _V:
NameError: name 'ReduceOp' is not defined
```

The use `TYPE_CHECKING` flag causes another error, since `allennlp.training.metrics.average` imports `ReduceOp` as a class object.

Reply:
Also, the dummy `ReduceOp` must have a dummy attribute `SUM`. It is used in `allennlp.training.metrics.average`.

Reply:
Ah, you're right, I was wrong about how the interpreter handles type hints. I apologize.

I still want to avoid making a dummy `ReduceOp`. So instead let's just remove the type hint from `dist_reduce()` and then in `allennlp.training.metrics.average` we could avoid directly importing `ReduceOp` and only use it when `is_distributed()` is `True`.

Reply:
I cannot rewrite `allennlp.training.metrics.average` since it is called from so many libraries. How about you @epwalsh?